### ISSUE TITLE
Unable to start Elastic Search using in-memory mode as configured

### ISSUE THREAD

**Issue by randall-sosacorp** [thinkaurelius/titan#700]
##### Expected result

Defining a configuration like `index.search.elasticsearch.local-mode` should start Elastic Search in embedded mode.
##### Actual result

The following exception is emitted:

```
org.elasticsearch.client.transport.NoNodeAvailableException: No node available
    at org.elasticsearch.client.transport.TransportClientNodesService.execute(TransportClientNodesService.java:219)
    at org.elasticsearch.client.transport.support.InternalTransportClusterAdminClient.execute(InternalTransportClusterAdminClient.java:85)
    at org.elasticsearch.client.support.AbstractClusterAdminClient.health(AbstractClusterAdminClient.java:123)
    at org.elasticsearch.action.admin.cluster.health.ClusterHealthRequestBuilder.doExecute(ClusterHealthRequestBuilder.java:93)
    at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:85)
    at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:59)
    at com.thinkaurelius.titan.diskstorage.es.ElasticSearchIndex.<init>(ElasticSearchIndex.java:198)
```
##### Affected versions

Titan 0.5.x, primarily 0.5.0-M2, 0.5.0-M3, and as of this writing, HEAD of the titan05 branch.
##### Detailed Description

This was discovered while running an integration test and attempting to initialize a Titan graph.  Debugging revealed that the `GraphDatabaseConfiguration` excludes "local-mode" configuration property because it is not defined globally within the INDEX_NS namespace.  This is manifested in the constructor of ElasticSearchIndex when the LOCAL_MODE `ConfigOption` evaluates to `false`.  The proposed fix is to change the LOCAL_MODE and CLIENT_ONLY `ConfigOption` from GLOBAL_OFFLINE to MASKABLE so that a local configuration file may take effect.

**Comment by dalaro** [comment]
@randall-sosacorp thanks for the cogent report and the PR complete with a test.  This is a priority for 0.5.0 final.  I the approach in #701, setting MASKABLE on the three concerned ES options, seems reasonable as a standalone change.  I also plan to look at the incident code in `GraphDatabaseConfiguration` to see what's going on, and in particular if this could be affecting other namespaces/options with lower mutability.

**Comment by dalaro** [comment]
@randall-sosacorp I pulled #701 into a branch and spent a while investigating.  The root cause of the failing test you submitted is actually completely unrelated to GraphDatabaseConfiguration.  GLOBAL_OFFLINE seems to be working as intended here.

The test fails because it has been written in such a way that the ElasticSearchIndex class which defines `client-only`, `local-mode`, etc. is never loaded.  Since the test avoids loading the ElasticSearchIndex class and its ConfigOption definitions, GraphDatabaseConfiguration perceives those settings as invalid and drops them.  This doesn't happen during ordinary `TitanFactory.open` use because the factory calls `ReflectiveConfigOptionLoader.loadOnce` to walk the classpath and force static ConfigOption initialization.  The test is kind of a special case that short-circuits this initialization.  When I force the static initializers in `ElasticSearchIndex` to run at the top of your test method, it passes.

In your deployment, did you initially start Titan with `client-only=true` & `local-mode=false` then later switch them?  Titan writes GLOBAL_OFFLINE config settings (among others) to its storage backend during its first startup.  Twiddling the config file value doesn't affect GLOBAL_OFFLINE properties after that first startup -- Titan only cares about the value in the storage backend, and those have to be changed through the `ManagementSystem` API by design.  Note that Rexster starting Titan counts, including Rexster started by titan.sh.  Starting in one config and then trying to use the other without going through `ManagementSystem` produces an exception that looks identical to the trace you posted.

As I replied earlier, I thought it would be useful for `local-mode` and `client-only` to be marked MASKABLE instead of GLOBAL_OFFLINE.  Now I don't see it.  It's only seemingly possible to switch between JVM-local discovery (ES only sees peers in the same process and doesn't use sockets) or using a round-robin TransportClient (ES opens network sockets to a list of hosts).  A heterogeneous mix of those two modes is not going to be useful since it seems like it would be necessarily split-brained.  Likewise, making `cluster-name` MASKABLE just seems like an invitation to accidentally split parts of one logical Titan index across two separate ES indices, leading to unexpected read behavior.  Please point out a use case if you see it.

I do think it would be useful to support a third mode in ElasticSearchIndex: instantiate a Node, but don't force `local(true)`.  That gets into potentially interesting use cases for MASKABLE, where all hosts run an ES Node, but some have `client(true)`/`data(false)` and basically just use their Node for intelligent request routing, but other hosts have live shards.  We don't have the plumbing for that yet though.

I'm closing this because I don't see anything to do right now, but please reopen/open new if you'd like to discuss further.

**Comment by randall-sosacorp** [comment]
tl;dr: Changing the configuration option to MASKABLE solved our issue.

The test associated with pull-request 701 was written so as to simulate what was observed during debugging.  A separate test I wrote that discovered this issue is not part of the pull request associated with this issue.

> This doesn't happen during ordinary TitanFactory.open use because the factory calls ReflectiveConfigOptionLoader.loadOnce to walk the classpath and force static ConfigOption initialization. 

I was able to observe that this, in fact, was not happening for our use-case.  When invoking `TitanFactory.open()`, Titan could not find an Elastic Search node when configured to run in embedded mode, which was the impetus for a debugging session.  Elastic Search did not run in "local-mode" because [this conditional](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-es/src/main/java/com/thinkaurelius/titan/diskstorage/es/ElasticSearchIndex.java#L128) evaluated to false.  This was because the configuration option could not be found in the `MergedConfiguration` via `config.get()`, which meant that its default value was used instead (false).  I endeavored to find out why it could not be discovered and hypothesized that it was [dropped in the block starting here](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/configuration/GraphDatabaseConfiguration.java#L1225).

> In your deployment, did you initially start Titan with client-only=true & local-mode=false then later switch them?

We did not, at least not wittingly.  Debugging revealed that our configuration remained unchanged within `GraphDatabaseConfiguration`, but once we entered `ElasticSearchIndex`, the configuration option could no longer be found.

> Likewise, making cluster-name MASKABLE just seems like an invitation to accidentally split parts of one logical Titan index across two separate ES indices, leading to unexpected read behavior.

Agreed.  This seems like a bad idea and was not necessary to solve our issue.  It's a change merely suggested because this configuration option seemed to suffer the same affliction, a first blush, as the "local-mode" option.

**Comment by randall-sosacorp** [comment]
I'd be remiss not to describe, in more detail, the debugging trace:
[TitanFactory.open()](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/core/TitanFactory.java#L92)
[Backend initialization](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/graphdb/configuration/GraphDatabaseConfiguration.java#L1604)
[Fetching indices from Backend](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/diskstorage/Backend.java#L131)
[Further down the stack](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/diskstorage/Backend.java#L412)
[Instantiating](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/diskstorage/Backend.java#L425)

This last line instantiates `ElasticSearchIndex`.  By this time, the configuration is passed in, missing the "local-mode" option.

**Comment by randall-sosacorp** [comment]
Nutshell summary: This was our problem, not Titan's.

After some further debugging, we were able to identify the root cause of the issue.  The [TitanFactory](https://github.com/thinkaurelius/titan/blob/0.5.0-M3/titan-core/src/main/java/com/thinkaurelius/titan/core/TitanFactory.java#L91) was not walking the classpath and forcing static `ConfigOption` initialization as expected.  The reason is that the `ReflectiveConfigOptionLoader` relies on the "reflections" library, which invokes this block of code:

``` java
    public static Set<URL> forJavaClassPath() {
        Set<URL> urls = Sets.newHashSet();

        String javaClassPath = System.getProperty("java.class.path");
        if (javaClassPath != null) {
            for (String path : javaClassPath.split(File.pathSeparator)) {
                try { urls.add(new File(path).toURI().toURL()); }
                catch (Exception e) { e.printStackTrace(); }
            }
        }
        return urls;
    }
```

relying on a system property to establish the classpath.  For the Play Framework, this is a bit tricky, especially when running in development mode.  The fix was to associate this system property with the correct value with respect to the Play Framework.

**Comment by dalaro** [comment]
@randall-sosacorp thank you for following up.  I've seen this behavior before with exec-maven-plugin's exec:java goal too.  I will look into potential alternatives to `ClasspathHelper.forJavaClassPath()` in #712 -- maybe there's a way to easily use the effective runtime classpath instead of the property?  I don't know yet but will look into it.

Related: #711
